src/hotspot/share/gc/g1/g1CardSetMemory.cpp

```diff
@@ -62,8 +62,8 @@ size_t G1CardSetAllocator::mem_size() const {
 
 size_t G1CardSetAllocator::wasted_mem_size() const {
   uint num_wasted_slots = _segmented_array.num_available_slots() -
-                          _segmented_array.num_allocated_slots() -
-                          (uint)_free_slots_list.pending_count();
+                         (_segmented_array.num_allocated_slots() -
+                          (uint)_free_slots_list.pending_count());
```

Comment:
I think the new calculation is wrong too. This is probably because of wrong documentation of `SegmentedArray::_num_available_slots` and `_num_allocated_slots`; in particular, `_num_allocated_slots` contains both free and pending counts.

So the correct solution would be (a bit abbreviated) I think:
```
unused_slots = (available - allocated) + _free_slots_list.free_count()
```
The first term calculates the not yet used slots, and the second directly adds the free slots.

Reply:
Thanks for the clarification. The new documentation on SegmentedArray::_num_available_slots and _num_allocated_slots makes sense to me.
Please correct me if I was wrong, so is the idea here to include free slots and exclude pending slots when counting unused slots?

Reply:
> I think the new calculation is wrong too. This is probably because of wrong documentation of `SegmentedArray::_num_available_slots` and `_num_allocated_slots`; in particular, `_num_allocated_slots` contains both free and pending counts.
> 
> So the correct solution would be (a bit abbreviated) I think:
> 
> ```
> unused_slots = (available - allocated) + _free_slots_list.free_count()
> ```
> 
> The first term calculates the not yet used slots, and the second directly adds the free slots.

Thanks for the clarification. The new documentation on SegmentedArray::_num_available_slots and _num_allocated_slots makes sense to me.
Please correct me if I was wrong, so is the idea here to include free slots and exclude pending slots when counting unused slots?

Reply:
> Please correct me if I was wrong, so is the idea here to include free slots and exclude pending slots when counting unused slots?

Yes. This has been a judgement call, there is no clear answer whether they should be counted as used or unused.

Reply:
> is the idea here to include free slots and exclude pending slots when counting unused slots?

I think not treating slots on pending lists as unused (or equivalently, treating slots on pending lists as "in-use") is cleaner. Since the free/pending counters are not strictly synchronized (`Atomic::add(&_free_count, count);` in `try_transfer_pending`), adding the pending counter in that formula could cause "unused slots > total slots", which is extremely confusing.

Reply:
These statistics are calculated during safepoint/gc, so there can be no synchronization issues.

Reply:
Got it. Thank you both for the explanation.